librz/core/rop.c

```diff
@@ -0,0 +1,1238 @@
+		RzListIter *iter;
+		RzCoreAsmHit *hit;
+		if (!core->analysis->ht_rop) {
+			core->analysis->ht_rop = ht_up_new(NULL, (HtUPFreeValue)rz_core_rop_gadget_info_free);
+		}
+		const ut64 addr_start = ((RzCoreAsmHit *)rz_list_first(hitlist))->addr;
+		rop_gadget_info = rz_core_rop_gadget_info_new(addr_start);
+		rz_list_foreach (hitlist, iter, hit) {
+			if (analyze_gadget(core, hit, rop_gadget_info) < 0) {
+				RZ_LOG_WARN("Failed to analyze gadget at 0x%" PFMT64x "\n", hit->addr);
+			}
+		}
+		ht_up_insert(core->analysis->ht_rop, addr_start, rop_gadget_info);
+	}
+	if (context->mask & RZ_ROP_GADGET_PRINT_DETAIL) {
+		print_rop_gadget_info(core, rop_gadget_info);
+	}
+	return 0;
+}
+
+static int fetch_search_itv(const RzCore *core, RzInterval *search_itv) {
+	rz_return_val_if_fail(core && core->config, -1);
+	const ut64 search_from = rz_config_get_i(core->config, "search.from"),
+		   search_to = rz_config_get_i(core->config, "search.to");
+	if (search_from > search_to && search_to) {
+		RZ_LOG_ERROR("core: search.from > search.to is not supported\n");
+		return -1;
+	}
+	search_itv->addr = search_from;
+	search_itv->size = search_to - search_from;
+
+	const bool empty_search_itv = search_from == search_to && search_from != UT64_MAX;
+	if (empty_search_itv) {
+		RZ_LOG_ERROR("core: `from` address is equal `to`\n");
+		return -1;
+	}
+	// TODO full address cannot be represented, shrink 1 byte to [0, UT64_MAX)
+	if (search_from == UT64_MAX && search_to == UT64_MAX) {
+		search_itv->addr = 0;
+		search_itv->size = UT64_MAX;
+	}
+	return 0;
+}
+
+static RzList /*<RzRopEndListPair *>*/ *compute_end_gadget_list(const RzCore *core, const ut8 *buf, const RzRopSearchContext *context) {
+	RzList /*<RzRopEndListPair *>*/ *end_list = rz_list_newf(free);
+	const int delta = context->to - context->from;
+
+	for (int i = 0; i < delta; i += context->increment) {
+		RzAnalysisOp end_gadget = RZ_EMPTY;
+		// Disassemble one.
+		rz_analysis_op_init(&end_gadget);
+		if (rz_analysis_op(core->analysis, &end_gadget, context->from + i, buf + i,
+			    delta - i, RZ_ANALYSIS_OP_MASK_BASIC) < 1) {
+			rz_analysis_op_fini(&end_gadget);
+			continue;
+		}
+
+		if (is_end_gadget(&end_gadget, context->crop)) {
+			RzRopEndListPair *epair = RZ_NEW0(RzRopEndListPair);
+			if (epair) {
+				epair->instr_offset = i + (end_gadget.delay ? context->increment : 0);
+				epair->delay_size = end_gadget.delay;
+				rz_list_append(end_list, epair);
+			}
+		}
+		rz_analysis_op_fini(&end_gadget);
+		if (rz_cons_is_breaked()) {
+			break;
+		}
+	}
+	return end_list;
+}
+
+static void set_increment_based_on_arch(const RzCore *core, const char *arch, int *increment) {
+	if (RZ_STR_EQ(arch, "mips")) { // MIPS has no jump-in-the-middle
+		*increment = 4;
+	} else if (RZ_STR_EQ(arch, "arm")) { // ARM has no jump-in-the-middle
+		*increment = rz_config_get_i(core->config, "asm.bits") == 16 ? 2 : 4;
+	} else if (RZ_STR_EQ(arch, "avr")) { // AVR is halfword aligned.
+		*increment = 2;
+	}
```

Comment:
@XVilka it would be nice to have an issue for this. seems that this should be an info that RzArch should be able to provide.